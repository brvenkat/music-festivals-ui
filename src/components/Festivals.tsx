import * as React from 'react'
import Card from '@material-ui/core/Card';
import { CardContent } from '@material-ui/core';
import { style } from 'typestyle'

export interface Festival {
  recordName: string
  bands: Band[]
}

interface Band {
  name: string
  festivals: string[]
}

const headerStyle = style({
  textAlign: 'center',
  fontWeight: 'bold',
  fontSize: '30px'
})

const cardStyle = style({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'space-between',
  marginBottom: '12px',
  color: 'blue',
  border: '2px solid blue'
})

export const Festivals = (props: { festivals: Festival[]}) => {
  const allFestivals = props.festivals
  return (
    <>
    <div className={headerStyle}>Music Festival Data</div>
    {
      allFestivals.map((festival, i) => (
          <Card key={i} className={cardStyle}>
            {festival.recordName}
            <CardContent>
              {
                festival.bands.map((band, j) => (
                  <span key={j}>
                    <p>{band.name}</p>
                    {
                      band.festivals.map((f, k) => (
                        <li key={k}>{f}</li>
                      ))
                    }
                  </span>
                ))
              }
            </CardContent>
          </Card>
      ))
    }
    </>
  )
}