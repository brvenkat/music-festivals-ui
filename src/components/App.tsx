
import * as React from 'react';
import { Festivals } from './Festivals'

export function App() {
  const [festivals, setFestivals] = React.useState([])
  const [loading, setLoading] = React.useState(true);

  React.useEffect(() => {
    fetch(process.env.MUSIC_FESTIVALS_URL)
    .then(res => res.json())
    .then(response => {
      setFestivals(response.festivals)
      setLoading(false)
    })
    .catch((error) => console.log(error))
  }, [])

  if (loading) {
    return (
      <div>Loading ...</div>
    )
  }
  if (!Array.isArray(festivals) || festivals.length === 0) {
    return (
      <div> Something has gone wrong, please refresh or try after some time</div>
    )
  }
  return <Festivals festivals={festivals} />
}
