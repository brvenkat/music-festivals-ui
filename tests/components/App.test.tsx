import * as React from 'react'
import { mount } from 'enzyme'
import { App } from '../../src/components/App'
import wait from 'waait'
import { Festivals } from '../../src/components/Festivals'

describe('App', () => {
  beforeEach(() => {
    window.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve) => {
          resolve({
              status: 200,
              ok: true,
              json: () => new Promise((resolve) => {
                  resolve({
                    festivals: [
                    {
                      recordName: 'record-1',
                      bands: [
                        {
                          name: 'band-1',
                          festivals: ['festival-1']
                        }
                      ]
                    }
                  ]});
              })
          });
      });
    })
  })
  it('Should create valid snapshot', () => {
    const wrapper = mount(<App />)
    expect(wrapper).toMatchSnapshot()
  })
  it('Should render all components correctly', async () => {
    const wrapper = mount(<App />)
    expect(window.fetch).toHaveBeenCalled()
    await wait()
    wrapper.update()
    expect(wrapper.find(Festivals).props().festivals).toEqual([
      {
        recordName: 'record-1',
        bands: [
          {
            name: 'band-1',
            festivals: ['festival-1']
          }
        ]
      }
    ])
  })
})