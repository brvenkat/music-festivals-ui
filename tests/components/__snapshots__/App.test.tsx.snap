// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should create valid snapshot 1`] = `ReactWrapper {}`;
